cinn/ir/ir_schedule_util.cc

```diff
@@ -789,84 +794,70 @@ void InsertBlock(Expr& for_loop, const Expr& insertion) {
+  return IterRange(new_min, new_extent);
 }
 
-std::vector<std::pair<Expr, Expr>> CalculateRequiredRegions(const Expr& block,
-                                                            const Expr& loop,
-                                                            const std::vector<Expr>& consumers,
-                                                            const Expr& root) {
+std::vector<IterRange> CalculateRequiredRegions(const Expr& block,
+                                                const Expr& loop,
+                                                const std::vector<Expr>& consumers,
+                                                const Expr& root) {
   CHECK(block.As<ir::ScheduleBlockRealize>());
   std::string block_tensor = GetTensor(block)->name;
-  std::vector<std::pair<Expr, Expr>> required_buffer_range;
+  std::vector<IterRange> required_buffer_range;
   CHECK(loop.As<ir::For>());
-  for (auto& i : consumers) {
-    CHECK(i.As<ir::ScheduleBlockRealize>());
-    Expr block_body = optim::IRCopy(i.As<ir::ScheduleBlockRealize>()->schedule_block.As<ir::ScheduleBlock>()->body);
-    auto iter_var   = i.As<ir::ScheduleBlockRealize>()->schedule_block.As<ir::ScheduleBlock>()->iter_vars;
-    auto iter_value = i.As<ir::ScheduleBlockRealize>()->iter_values;
-    ReplaceExpr(&block_body, iter_var, iter_value);
+  // deduce accessed regions of the tensor in block by itering each consumer
+  for (const auto& load_block : consumers) {
+    CHECK(load_block.As<ir::ScheduleBlockRealize>());
+    Expr block_body =
+        optim::IRCopy(load_block.As<ir::ScheduleBlockRealize>()->schedule_block.As<ir::ScheduleBlock>()->body);
+    auto iter_vars   = load_block.As<ir::ScheduleBlockRealize>()->schedule_block.As<ir::ScheduleBlock>()->iter_vars;
+    auto iter_values = load_block.As<ir::ScheduleBlockRealize>()->iter_values;
+    ReplaceExpr(&block_body, iter_vars, iter_values);
     // Notice that we look for For nodes in loop's body instead of loop itself.
     auto find_loops = ir::CollectIRNodesWithoutTensor(
-        loop.As<ir::For>()->body, [&](const Expr* x) { return x->As<ir::For>() && Contains(*x, i); });
-    auto find_load = ir::CollectIRNodesWithoutTensor(block_body, [&](const Expr* x) {
+        loop.As<ir::For>()->body, [&](const Expr* x) { return x->As<ir::For>() && Contains(*x, load_block); });
+    auto load_nodes = ir::CollectIRNodesWithoutTensor(block_body, [&](const Expr* x) {
       return x->As<ir::Load>() && x->As<ir::Load>()->tensor.as_tensor_ref()->name == block_tensor;
     });
-    for (auto& load : find_load) {
+
+    // collect vars and their ranges of each loop under the input loop
+    std::vector<Var> loop_vars;
+    std::vector<IterRange> loop_ranges;
+    for (const auto& for_loop : find_loops) {
+      loop_vars.emplace_back(for_loop.As<ir::For>()->loop_var);
+      loop_ranges.emplace_back(for_loop.As<ir::For>()->min, for_loop.As<ir::For>()->extent);
```

Comment:
`emplace_back(IterRange(...));`

to increase readability? I saw some other places use this techniques. Usually, for big scale project, we prevent implicit conversion to avoid loss of Type information during reading.

You could check the reference to read and learn:
1. Google coding style, the pros and cons of implicit conversion, implicit conversion is not always a bad thing, but learn the cons and pros when you use them
https://google.github.io/styleguide/cppguide.html#Implicit_Conversions

2. More Effective C++
Search text "In most cases, the inconvenience of having to call conversion functions explicitly is more than compensated for by the fact that unintended functions can no longer be silently invoked" in `More Effective C++`. C++ committee also tried to avoid implicit conversion in most cases.
http://ptgmedia.pearsoncmg.com/imprint_downloads/informit/aw/meyerscddemo/DEMO/MEC/MC2_FR.HTM#p25

Reply:
actually no implicit conversion happens here, it just call the `IterRange` constructor with arguments by `template< class... Args > void emplace_back( Args&&... args )`. Anyway, I take your advice for readability.

Reply:
There is same readability / implicit conversion reason behind implicit conversion and implicit constructor ...

Reply:
^^